test/foundry/unit/TokenDistributor.t.sol

```diff
@@ -0,0 +1,1446 @@
+import {TokenDistributorInternals} from "../utils/TokenDistributorInternals.sol";
+import {Kwenta} from "../../../contracts/Kwenta.sol";
+import {StakingRewardsV2} from "../../../contracts/StakingRewardsV2.sol";
+import {RewardEscrowV2} from "../../../contracts/RewardEscrowV2.sol";
+import {StakingSetup} from "../utils/setup/StakingSetup.t.sol";
+
+contract TokenDistributorTest is StakingSetup {
+    event CheckpointToken(uint time, uint tokens);
+    event EpochClaim(address user, uint epoch, uint tokens);
+
+    TokenDistributor public tokenDistributor;
+
+    function setUp() public override {
+        /// @dev starts after a week so the startTime is != 0
+        goForward(1 weeks + 1);
+        super.setUp();
+        switchToStakingV2();
+        tokenDistributor = new TokenDistributor(
+            address(kwenta),
+            address(stakingRewardsV2),
+            address(rewardEscrowV2),
+            0
+        );
+        vm.prank(treasury);
+        kwenta.transfer(address(this), 100_000 ether);
+    }
+
+    /// @notice checkpointToken happy case after 1 week
+    function testCheckpointToken() public {
+        kwenta.transfer(address(tokenDistributor), 10);
+        kwenta.transfer(address(user1), 1);
+        vm.startPrank(address(user1));
+        kwenta.approve(address(stakingRewardsV2), 1);
+        stakingRewardsV2.stake(1);
+        goForward(1 weeks);
+
+        vm.expectEmit(true, true, true, true);
+        emit CheckpointToken(2 weeks + 2, 10);
+        tokenDistributor.checkpointToken();
+    }
+
+    /// @notice checkpointToken for missed weeks
+    function testCheckpointTokenManyMissed() public {
+        kwenta.transfer(address(tokenDistributor), 10);
+        goForward(5 weeks);
+        vm.expectEmit(true, true, true, true);
+        emit CheckpointToken(6 weeks + 2, 10);
+        tokenDistributor.checkpointToken();
+    }
+
+    /// @notice checkpointToken for sinceLast == 0
+    function testManyCheckpointTokenAtOnce() public {
+        kwenta.transfer(address(tokenDistributor), 10);
+        kwenta.transfer(address(user1), 1);
+        vm.startPrank(address(user1));
+        kwenta.approve(address(stakingRewardsV2), 1);
+        stakingRewardsV2.stake(1);
+        goForward(1 weeks);
+
+        vm.expectEmit(true, true, true, true);
+        emit CheckpointToken(2 weeks + 2, 10);
+        tokenDistributor.checkpointToken();
+        vm.expectEmit(true, true, true, true);
+        emit CheckpointToken(2 weeks + 2, 0);
+        tokenDistributor.checkpointToken();
+    }
+
+    /// @notice checkpoint at the start and < 1 week
+    /// make sure theres no error dividing by 0
+    /// because thisEpoch should be 0
+    function testCheckpointTokenFirstWeek() public {
+        tokenDistributor.checkpointToken();
+        goForward(1 days);
+        tokenDistributor.checkpointToken();
+    }
+
+    /// @notice claimEpoch happy case
+    function testClaimEpoch() public {
```

Comment:
Missing a tests for > 1 person claiming and making sure each person gets expected amount